### DISCUSSION TITLE
Extend HiveTypeParser to support Presto types

### DISCUSSION THREAD

**Discussion by majetideepak** [facebookincubator/velox#7502] [General]
We have a [HiveTypeParser](https://github.com/facebookincubator/velox/blob/main/velox/type/fbhive/HiveTypeParser.h) in Velox to parse Hive Types. I see that this is mainly used in testing (DWRF mostly) and in RemoteFunctionService.

Why do we use Hive Types style for testing? Why not Presto style?

We use antlr to parse Presto Type in Prestissimo. There is now a need to move the Presto Type parser to Velox to support PrestoQueryRunner in Velox.

We can easily extend the HiveTypeParser to parse Presto Types.
The missing pieces are 
1) Start/end sub types literal is `<` in Hive and `(` in Presto. We can make this configurable.
2) Hive Types use a semi-colon between name and type. Presto does not. We can make this configurable.
3) Missing parse support for `functiontype`. Should be straightforward to add this.
4) Allow optional field names for struct.
5) Missing support for types with spaces such as `timestamp with timezone`.
6) Better error handling.

We can remove the dependency on antlr in Prestissimo with this support.

**Comment by mbasmanova** [comment]
HiveTypeParser is a legacy code and we shouldn't be using it. Tests can create types more easily using macros: ROW, ARRAY, MAP, etc.

**Comment by mbasmanova** [comment]
There is also type parsing logic in velox/expression/FunctionSignature.cpp

```
TypeSignature parseTypeSignature(const std::string& signature) {
```

**Reply by mbasmanova** [reply]
@pedroerp Pedro is extending parseTypeSignature in https://github.com/facebookincubator/velox/pull/7523

**Comment by duanmeng** [comment]
> There is also type parsing logic in velox/expression/FunctionSignature.cpp
> 
> ```
> TypeSignature parseTypeSignature(const std::string& signature) {
> ```

@mbasmanova @majetideepak I used to want to reuse this method and add an API to transform the `TypeSignature` to Velox Type, something like the following code, this could be a temporary solution to port the PrestoQueryRunner to Velox, but it is kind of hardcode. Adding a type parser using flex/bison is more elegant. 

```C++
namespace {
TypePtr toVeloxType(const TypeSignature& typeSignature) {
  if (typeSignature.baseName() == "bigint") {
    return BIGINT();
  } else if (typeSignature.baseName() == "double") {
    return DOUBLE();
  } else if (typeSignature.baseName() == "map") {
    auto keyType = toVeloxType(typeSignature.parameters().front());
    auto valType = toVeloxType(typeSignature.parameters().back());
    return MAP(keyType, valType);
  } else if (typeSignature.baseName() == "array") {
    auto elementType = toVeloxType(typeSignature.parameters().front());
    return ARRAY(elementType);
  } else { // row
    std::vector<TypePtr> elementTypes;
    for (const auto& typeName : typeSignature.parameters()) {
      elementTypes.emplace_back(toVeloxType(typeName));
    }
    return ROW(std::move(elementTypes));
  }
}
} // namespace

TEST(ParseTypeSignatureTest, debug) {
  ASSERT_NE(INTEGER(), toVeloxType(parseTypeSignature("bigint")));
  ASSERT_EQ(BIGINT(), toVeloxType(parseTypeSignature("bigint")));
  ASSERT_EQ(DOUBLE(), toVeloxType(parseTypeSignature("double")));
  ASSERT_TRUE(MAP(BIGINT(), DOUBLE())->equivalent(
      *toVeloxType(parseTypeSignature("map(bigint, double)"))));
  ASSERT_TRUE(MAP(BIGINT(), ARRAY(DOUBLE()))->equivalent(
      *toVeloxType(parseTypeSignature("map(bigint, array(double))"))));
  ASSERT_TRUE(ROW({BIGINT(), MAP(BIGINT(), ARRAY(DOUBLE()))})->equivalent(
      *toVeloxType(parseTypeSignature("row(bigint,map(bigint, array(double)))"))));
}

```

**Comment by mbasmanova** [comment]
> I used to want to reuse this method and add an API to transform the TypeSignature to Velox Type, something like the following code, this could be a temporary solution to port the PrestoQueryRunner to Velox

@duanmeng Sounds like we can use this to get unblocked. It would be nice to continue making progress on brining PrestoQueryRunner to Velox and starting to run Fuzzer using that.

**Reply by duanmeng** [reply]
@mbasmanova OK, I've created a PR https://github.com/facebookincubator/velox/pull/7561. cc @majetideepak

**Reply by majetideepak** [reply]
I have a draft PR for the parser using Flex/Bison here as well https://github.com/facebookincubator/velox/pull/7568
All the tests from the Prestissimo antlr implementation `presto-native-execution/presto_cpp/main/types/tests/TypeSignatureTest.cpp` pass sans a difference in the error message.

**Reply by duanmeng** [reply]
OK, I will close #7561 and start porting the PrestoQueryRunner simultaneously.

**Comment by majetideepak** [comment]
This discussion is now resolved with https://github.com/facebookincubator/velox/pull/7568
